### PR TITLE
Fix SQL error detection problems.

### PR DESCRIPTION
**Pull request by ruschein**

### CODE REVIEW

File: cpp/lib/src/DbConnection.cc [hunk 1]
```diff
@@ -127,6 +127,37 @@ int DbConnection::getLastErrorCode() const {
 }
 
 
+static std::vector<std::string> SplitMySQLStatements(const std::string &query) {
+    std::vector<std::string> statements;
+
+    std::string statement;
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:133]: should be inside the for loop

File: cpp/lib/src/DbConnection.cc [hunk 2]
```diff
@@ -127,6 +127,37 @@ int DbConnection::getLastErrorCode() const {
...
+        } else if (unlikely(ch == ';')) {
+            statements.emplace_back(statement);
+            statement.clear();
+        } else {
+            if (ch == '\'' or ch == '"')
+                current_quote = ch;
+            statement += ch;
+        }
+    }
+    if (not statement.empty())
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:154]: should be inside the for loop

File: cpp/lib/src/DbConnection.cc [hunk 3]
```diff
@@ -127,6 +127,37 @@ int DbConnection::getLastErrorCode() const {
...
+                current_quote = '\0';
+            else if (ch == '\\')
+                escaped = true;
+        } else if (unlikely(ch == ';')) {
+            statements.emplace_back(statement);
+            statement.clear();
+        } else {
+            if (ch == '\'' or ch == '"')
+                current_quote = ch;
+            statement += ch;
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:151]: What about MySQL comments?
https://dev.mysql.com/doc/refman/8.0/en/comments.html

File: cpp/lib/src/DbConnection.cc (hunk 1 above)
Comment (mtrojan-ub) [reply cpp/lib/src/DbConnection.cc:133]: hmm or maybe not

File: cpp/lib/src/DbConnection.cc (hunk 2 above)
Comment (mtrojan-ub) [reply cpp/lib/src/DbConnection.cc:154]: hmm or maybe not

File: cpp/lib/src/DbConnection.cc (hunk 1 above)
Comment (ruschein) [reply cpp/lib/src/DbConnection.cc:133]: Why?

File: cpp/lib/src/DbConnection.cc (hunk 2 above)
Comment (ruschein) [reply cpp/lib/src/DbConnection.cc:154]: Nope.

File: cpp/lib/src/DbConnection.cc [hunk 4]
```diff
@@ -127,6 +127,37 @@ int DbConnection::getLastErrorCode() const {
...
+static std::vector<std::string> SplitMySQLStatements(const std::string &query) {
+    std::vector<std::string> statements;
+
+    std::string statement;
+    char current_quote('\0'); // NUL means not currently in a string constant.
+    bool escaped(false); // backslash not yet seen
+    for (const char ch : query) {
+        if (current_quote != '\0') {
+            statement += ch;
+            if (escaped)
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:139]: I thought escaping is only possible when inside a string constant, not outside

Comment (mtrojan-ub) [reply cpp/lib/src/DbConnection.cc:139]: ah right.. sorry my brain hurts

File: cpp/lib/src/DbConnection.cc [hunk 5]
```diff
@@ -127,6 +127,37 @@ int DbConnection::getLastErrorCode() const {
...
+                escaped = false;
+            else if (ch == current_quote)
+                current_quote = '\0';
+            else if (ch == '\\')
+                escaped = true;
+        } else if (unlikely(ch == ';')) {
+            statements.emplace_back(statement);
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:146]: Trim before emplacing?

File: cpp/lib/src/DbConnection.cc [hunk 6]
```diff
@@ -271,11 +305,9 @@ bool DbConnection::queryFile(const std::string &filename) {
         return false;
 
     if (type_ == T_MYSQL) {
-        ::mysql_set_server_option(&mysql_, MYSQL_OPTION_MULTI_STATEMENTS_ON);
         const bool query_result(query(StringUtil::TrimWhite(&statements)));
         if (query_result)
             mySQLSyncMultipleResults();
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:310]: If query handles everything we do not need to sync all results anymore here

File: cpp/lib/src/DbConnection.cc [hunk 7]
```diff
@@ -127,6 +127,58 @@ int DbConnection::getLastErrorCode() const {
 }
 
 
+enum CommentFlavour { NO_COMMENT, C_STYLE_COMMENT, END_OF_LINE_COMMENT };
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:130]: C_STYLE_COMMENT is never used

File: cpp/lib/src/DbConnection.cc [hunk 8]
```diff
@@ -127,6 +127,58 @@ int DbConnection::getLastErrorCode() const {
...
+                escaped = true;
+        } else if (unlikely(*ch == ';')) {
+            statements.emplace_back(statement);
+            statement.clear();
+        } else if (unlikely(*ch == '#'))
+            comment_flavour = END_OF_LINE_COMMENT;
+        else {
+            // Check for comments starting with "/*" or "-- ":
+            if ((*ch == '/' and ch + 1 < query.cend() and *(ch + 1) == '*')
+                or (*ch == '-' and ch + 2 < query.cend() and *(ch + 1) == '-' and *(ch + 2) == ' '))
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:161]: END_OF_LINE_COMMENT should be set as well if "-- "

File: cpp/lib/src/DbConnection.cc [hunk 9]
```diff
@@ -127,6 +127,61 @@ int DbConnection::getLastErrorCode() const {
...
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:161]: move the second line into an elseif and you dont need the other if...else branches

File: cpp/lib/src/DbConnection.cc [hunk 10]
```diff
@@ -127,6 +127,61 @@ int DbConnection::getLastErrorCode() const {
...
+                    ++ch;
+                    comment_flavour = C_STYLE_COMMENT;
+                } else {
+                    ch += 2;
+                    comment_flavour = END_OF_LINE_COMMENT;
+                }
+                continue;
+            }
+
+            if (*ch == '\'' or *ch == '"')
```
Comment (mtrojan-ub) [inline cpp/lib/src/DbConnection.cc:173]: ... and this into else

**Pull request merged**